### PR TITLE
[Validator] Choices constraint improvement  

### PR DESCRIPTION
**Pull request by nikophil**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no 
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT

Hi,

here is a little improvement for the choice constraint:  expose a new `choices` wildcard for the messages, in order to provide a way to display easily the valid choices to the user.

thanks.

### CODE REVIEW

**Comment by HeahDude** [comment]
This can be closed as this feature is supported by the annotation reader already, just remove the quotes :).

```php
 /**
     * @var string
     *
     * @Assert\Choice(choicesAsConstant=App\Constants::AVAILABLE_STATES)
     */
    private $state;
```

**Comment by yceruto** [comment]
@HeahDude Do you meant `@Assert\Choice(choices=App\Constants::AVAILABLE_STATES)`?

**Comment by HeahDude** [comment]
yes x).

**Comment by nikophil** [comment]
yeah indeed, you're right, and i 've learnt something new :)
will we keep the new wildcard `choices` i've added, or should i close the PR ?

**Comment by HeahDude** [comment]
I'm 👍 for supporting a default choices implode, as long as we don't change the default message for both BC and let it opt-in performance wise.

**Comment by nikophil** [comment]
i've updated the PR

File: src/Symfony/Component/Validator/Constraints/ChoiceValidator.php [hunk 1]
```diff
@@ -100,6 +100,7 @@ public function validate($value, Constraint $constraint)
         } elseif (!\in_array($value, $choices, true)) {
             $this->context->buildViolation($constraint->message)
                 ->setParameter('{{ value }}', $this->formatValue($value))
+                ->setParameter('{{ choices }}', implode(', ', $choices))
```
Comment (ro0NL) [inline src/Symfony/Component/Validator/Constraints/ChoiceValidator.php]: should also be added to the multiple=true case

choices can be any value,  i think we need to map all choices using `formatValue()` also.

Comment (nikophil) [reply src/Symfony/Component/Validator/Constraints/ChoiceValidator.php]: yep, you're totally right - i've added to the multiple=true case and formatted the value

**Comment by xabbuh** [comment]
@nikophil Can you please update the PR title and description to match the recent changes (they will be part of the git history)?

**Comment by nikophil** [comment]
done :+1:

**Review by xabbuh — APPROVED** [review]

**Review by HeahDude — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nikophil.

**Pull request merged**